internal/component/loki/source/file/decompresser.go

```diff
@@ -310,26 +309,22 @@ func (d *decompressor) markPositionAndSize() error {
 }
 
 func (d *decompressor) Stop() {
-	d.mut.Lock()
-	d.stopping = true
-	defer func() {
-		d.stopping = false
-	}()
-	d.mut.Unlock()
```

Comment:
I dont get why we did it like this here. If this gets called it it is from [here](https://github.com/grafana/alloy/blob/main/internal/component/loki/source/file/runner.go#L60). So the task is done anyway but maybe I am missing something

Reply:
Yes it should have locked the mutext when chaning the state. 

>Yes, it looks like in internal/component/loki/source/file/runner.go in the for loop r.reader.Run() could keep getting called until the backoff stops.

While this is true it never happens after stop is called because then the task exits.

Reply:
Just a bit more context. `tailer`  and `decompressor` behaves a bit different.

When calling `Run` on a `tailer` it will start tailing the file and only exit when the job quits. The loop you mention will only happen if we encounter any errors during setup.

When calling `Run` on a `decompressor` we read and decompress the file and go through all lines from the position we have saved and then exit, but then that job will be retried again even though we have finished the file. 

So we don't want to spawn a go routine for `updatePosition` ever time we call run, just the first time. All other calls just makes sure we have a done channel set and read / tails the file. 

Stopped is only called when the task is canceled so we are never calling stop after that.